src/v1/parser.rs

```diff
@@ -0,0 +1,814 @@
+
+    use ::HashType;
+    use super::{Entry, Footer, Hashes, Header, ParseRowError};
+    use super::{parse_hex, is_hex, is_hex_encoding, unescape_hex};
+
+    #[test]
+    fn test_header_parse() {
+        let res = Header::parse(b"");
+        assert!(matches!(res,
+                Err(ParseRowError(ref msg))
+                if msg.starts_with("Invalid signature:")),
+            "Result was: {:?}", res);
+
+        let res = Header::parse(b"\xff");
+        assert!(matches!(res,
+                Err(ParseRowError(ref msg))
+                if msg.starts_with("Expected utf-8 string:")),
+            "Result was: {:?}", res);
+
+        let res = Header::parse(b"DIRSIGNATURE");
+        assert!(matches!(res,
+                Err(ParseRowError(ref msg))
+                if msg.starts_with("Missing version")),
+            "Result was: {:?}", res);
+
+        let res = Header::parse(b"DIRSIGNATURE.v2");
+        assert!(matches!(res,
+                Err(ParseRowError(ref msg))
+                if msg.starts_with("Invalid version:")),
+            "Result was: {:?}", res);
+
+        let res = Header::parse(b"DIRSIGNATURE.v1");
+        assert!(matches!(res,
+                Err(ParseRowError(ref msg))
+                if msg.starts_with("Missing hash type")),
+            "Result was: {:?}", res);
+
+        let res = Header::parse(b"DIRSIGNATURE.v1 sha512/25");
+        assert!(matches!(res,
+                Err(ParseRowError(ref msg))
+                if msg.starts_with("Invalid hash type: Unsupported hash algorithm")),
+            "Result was: {:?}", res);
+
+        let res = Header::parse(b"DIRSIGNATURE.v1 sha512/256 size=2");
+        assert!(matches!(res,
+                Err(ParseRowError(ref msg))
+                if msg.starts_with("Expected block_size attribute")),
+            "Result was: {:?}", res);
+
+        let res = Header::parse(b"DIRSIGNATURE.v1 sha512/256 block_size=dead");
+        assert!(matches!(res,
+                Err(ParseRowError(ref msg))
+                if msg.starts_with("Cannot parse block size \"dead\":")),
+            "Result was: {:?}", res);
+
+        let res = Header::parse(b"DIRSIGNATURE.v1 sha512/256 block_size=1234");
+        let header = res.unwrap();
+        assert_eq!(header.get_version(), "v1");
+        assert!(matches!(header.get_hash_type(), HashType::Sha512_256));
+        assert_eq!(header.get_block_size(), 1234);
+    }
+
+    #[test]
+    fn test_entry_parse() {
+        let t = HashType::Sha512_256;
+        let b = 32768;
+
+        let res = Entry::parse(b"", Path::new(""), t, b);
+        assert!(matches!(res, Ok(None)));
+
+        let res = Entry::parse(b"/test", Path::new("/dir"), t, b);
+        assert!(matches!(res,
+                Ok(Some(Entry::Dir(ref dir_path)))
+                if dir_path == Path::new("/test")),
+            "Result was: {:?}", res);
+
+        // FIXME: IMHO next test should pass
+        // TODO: writer also should double backslashes
```

Comment:
f45886a4473fbcde48897b9170027c6b82259447 fixes the escaping issue (and the spec)